File: Mixter.Domain/Core/Subscriptions/Subscription.cs
using System.Collections.Generic;
using Mixter.Domain.Core.Messages;
using Mixter.Domain.Core.Subscriptions.Events;
using Mixter.Domain.Identity;

namespace Mixter.Domain.Core.Subscriptions
{
    public class Subscription
    {
        private readonly DecisionProjection _projection;

        public Subscription(IEnumerable<IDomainEvent> events)
        {
            _projection = new DecisionProjection();

            foreach (var evt in events)
            {
                _projection.Apply(evt);
            }
        }

        public static void FollowUser(IEventPublisher eventPublisher, UserId follower, UserId followee)
        {
            var userFollowed = new UserFollowed(new SubscriptionId(follower, followee));
            eventPublisher.Publish(userFollowed);
        }

        public void Unfollow(IEventPublisher eventPublisher)
        {
            eventPublisher.Publish(new UserUnfollowed(_projection.Id));
        }

        public void NotifyFollower(IEventPublisher eventPublisher, MessageId messageId)
        {
            if (_projection.IsActive)
            {
                eventPublisher.Publish(new FolloweeMessageQuacked(_projection.Id, messageId));
            }
        }

        private class DecisionProjection : DecisionProjectionBase
        {
            public DecisionProjection()
            {
                AddHandler<UserFollowed>(When);
                AddHandler<UserUnfollowed>(When);
            }

            public SubscriptionId Id { get; private set; }

            public bool IsActive { get; private set; }

            private void When(UserFollowed evt)
            {
                Id = evt.SubscriptionId;
                IsActive = true;
            }

            private void When(UserUnfollowed evt)
            {
                IsActive = false;
            }
        }
    }
}

File: domain/src/main/java/mixter/domain/core/subscription/Subscription.java
package mixter.domain.core.subscription;

import mixter.doc.Aggregate;
import mixter.domain.DecisionProjectionBase;
import mixter.domain.Event;
import mixter.domain.EventPublisher;
import mixter.domain.core.subscription.events.UserFollowed;
import mixter.domain.core.subscription.events.UserUnfollowed;
import mixter.domain.identity.UserId;

import java.util.List;

@Aggregate
public class Subscription {
    private DecisionProjection projection;

    public Subscription(List<Event> history) {
        projection = new DecisionProjection(history);
    }

    public static void follow(UserId follower, UserId followee, EventPublisher eventPublisher) {
        eventPublisher.publish(new UserFollowed(new SubscriptionId(follower, followee)));
    }

    public void unfollow(EventPublisher eventPublisher) {
        eventPublisher.publish(new UserUnfollowed(projection.id));
    }

    private class DecisionProjection extends DecisionProjectionBase {
        public SubscriptionId id;

        public DecisionProjection(List<Event> history) {
            super();
            super.register(UserFollowed.class, this::apply);
            history.forEach(this::apply);
        }

        private void apply(UserFollowed event) {
            id = event.getSubscriptionId();
        }
    }
}